fc-extract-delta/main.go

```diff
@@ -0,0 +1,178 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+	"strings"
+
+	manet "github.com/multiformats/go-multiaddr-net"
+	"github.com/multiformats/go-multiaddr"
+	"github.com/filecoin-project/lotus/api/client"
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/urfave/cli/v2"
+	"github.com/ipfs/go-cid"
+
+	"github.com/filecoin-project/oni/fc-extract-delta/lib"
+)
+
+var (
+	fromFlag = cli.StringFlag{
+		Name: "from",
+		Usage: "block CID of initial state",
+		Required: true,
+	}
+
+	toFlag = cli.StringFlag{
+		Name: "to",
+		Usage: "block CID of ending state",
+		Required: true,
+	}
+
+	apiFlag = cli.StringFlag{
+		Name: "api",
+		Usage: "api endpoint, formatted as token:multiaddr",
+		Value: "",
+		EnvVars: []string{"FULLNODE_API_INFO"},
+	}
+	
+	idFlag = cli.StringFlag{
+		Name: "cid",
+		Usage: "CID to act upon",
+		Required: true,
+	}
+)
+
+var deltaCmd = &cli.Command{
+	Name: "delta",
+	Description: "Collect affected state between two tipsets",
+	Flags: []cli.Flag{&fromFlag,&toFlag, &apiFlag},
+	Action: extract,
+}
+
+var messageCmd = &cli.Command{
+	Name: "message",
+	Description: "Extract affected actors from a single message",
+	Flags: []cli.Flag{&idFlag, &apiFlag},
+	Action: message,
+}
+
+func makeClient(api string) (api.FullNode, error) {
+	sp := strings.SplitN(api, ":", 2)
+	if len(sp) != 2 {
+		return nil, fmt.Errorf("invalid api value, missing token or address: %s", api)
+	}
+	//TODO: discovery from filesystem
+	token := sp[0]
+	ma, err := multiaddr.NewMultiaddr(sp[1])
+	if err != nil {
+		return nil, fmt.Errorf("could not parse provided multiaddr: %w", err)
+	}
+	_, dialAddr, err := manet.DialArgs(ma)
+	if err != nil {
+		return nil, fmt.Errorf("invalid api multiAddr: %w", err)
+	}
+	addr := "ws://" + dialAddr + "/rpc/v0"
+
+	headers := http.Header{}
+	if len(token) != 0 {
+		headers.Add("Authorization", "Bearer "+string(token))
+	}
+	node, _, err := client.NewFullNodeRPC(addr, headers)
+	if err != nil {
+		return nil, fmt.Errorf("could not connect to api: %w", err)
+	}
+	return node, nil
+}
```

Comment:
Heads up: at https://github.com/filecoin-project/oni/pull/167/files#diff-3fc1d1cd75e2088d9da8d680c52a0458R30-R63 I am using directly a `repo` and `BlockStore` and `ChainStore` - I noticed that most of the API methods related to the blockchain are pretty much forwarded to the `ChainStore`, so I haven't had a need for `api.FullNode` so far.

This makes it easier to just point to the Lotus repo directory, even if the node is not running.

Reply:
seems easy enough to abstract out. the benefit of `api.FullNode` is that interface is the thing that should be consistent for other implementatiuons, while the on-disk format of `BlockStore`/`ChainStore` may be lotus specific.